flink-core/src/main/java/org/apache/flink/api/common/io/OutputFormatBase.java

```diff
@@ -97,21 +108,35 @@ private void tryAcquire(int permits) throws IOException {
         }
     }
 
+    /**
+     * Asynchronously write a record and deal with {@link OutputFormatBase#maxConcurrentRequests}.
+     * To specify how a record is written, please override the {@link OutputFormatBase#send(Object)}
+     * method.
+     */
     @Override
-    public void writeRecord(OUT record) throws IOException {
+    public final void writeRecord(OUT record) throws IOException {
         checkAsyncErrors();
         tryAcquire(1);
-        final ListenableFuture<V> result;
+        final CompletionStage<V> result;
         try {
             result = send(record);
         } catch (Throwable e) {
             semaphore.release();
             throw e;
         }
-        Futures.addCallback(result, callback);
+        Futures.addCallback(
+                completableFutureToListenableFuture(result.toCompletableFuture()),
+                callback,
+                Executors.directExecutor());
```

Comment:
This is overly complicated.

When you have a `CompletionStage` you can just apply a `whenComplete`. with the callback being re-written to a `BiConsumer<V, Throwable>`.

Reply:
This will also remove the need for the odd `send()` javadoc that `toCompletableFuture()` must be properly implemented. (Which is weird in any case since , you know, every method of an interface should be properly implemented.) Realistically this will just be a CompletableFuture anyway.

Reply:
> This will also remove the need for the odd `send()` javadoc that `toCompletableFuture()` must be properly implemented. (Which is weird in any case since , you know, every method of an interface should be properly implemented.) Realistically this will just be a CompletableFuture anyway.

I agree, the javadoc precision is useless because redundant.

Reply:
>I know but using whenComplete won't work as get() is never called on the result future

Sure it does. `whenComplete` is run once the future is completed, not when the result is requested.

Reply:
Fixed

Reply:
Fixed